### DISCUSSION TITLE
[非官方] 主页标准与指导

### DISCUSSION THREAD

**Discussion by Light-Beacon** [Meloong-Git/PCL#4483] [自定义主页]
# 前言
作为多年老牌主页——新闻主页的作者，我是欢迎大家创建并提交自己的主页的，在 [#2528](https://github.com/Hex-Dragon/PCL2/discussions/2528) 也看了很多大家创建的主页，很高兴 PCL 社区能有这样积极的一群内容创作者。但就目前看来提交的主页质量参差不齐，大部分都有其相应的问题，想到头来也许缺乏相应的经验和标准吧，也就是没有人告诉应该做什么，不应该做什么。于是结合之前的问题，我有感而生了本文，希望能帮助主页创作者们创造出质量更好的主页。

>[!IMPORTANT]
> 注意：这里不是官方硬性要求，本文也不代表 PCL 官方发出，主页硬性要求见 [#2528](https://github.com/Hex-Dragon/PCL2/discussions/2528)，这里只是我**个人**推荐大家这么做。

如果本文有疏漏或有建议可以在下方回复。内容将来有可能会扩充 ~~（想到啥写啥）~~

<details>
<summary>【附】第一次接触联网自定义主页需要注意的事 </summary>
如果你是第一次接触联网自定义主页，提前说一下自定义联网主页有两个你必须要迈过去的技术门槛：

* 你需要了解并会使用 XAML 语言
* 你需要对网络有一定理解，并拥有一套稳定的网络服务
    * 了解 HTTP ，因为 PCL 使用 HTTP 协议获取主页
    * 拥有/租用 服务器或内容分发网络等网络设施
    * 若不使用第三方对象存储，应熟悉使用 web 服务器软件（如 Nginx 等）
    * 拥有自己的域名（虽然没有应该也可以），会设置解析

看着花里胡哨的，其实都并不是很难。对于第一条，如果你有一些经验的话你可以使用 VS 构建一个 WPF 应用往里拖控件然后看看代码啥的，应该很快就能上手。第二条的话不会相关技术这里真的建议去学习一下，毕竟这个还是蛮重要的基础（也很简单）。
</details>

---

# 主页推荐标准与指导
### 内容
* 应遵守 [《用户协议与免责声明》](https://shimo.im/docs/rGrd8pY8xWkt6ryW#anchor-i7r2) 的相关条目
* *不包含服务器、第三方网站的广告性质链接，间接广告也不行* （这里不建议包含任何性质的服务器链接）
* 主页一定要拥有自己独特的功能：如展示MC新闻、历史上的今天等等
    * 不要制作各种功能的快捷入口的这类主页
        * ~~除非你真的有信心能甩简单主页几条街~~
    * 可以添加一些辅助性的按钮，如 “清理内存” 等，但不应过多，整体不应超过整个主页内容的 20%
* 应当在适当位置添加反馈入口

### 布局
* 无论如何缩放（过大除外），尽量使你的主页无论在哪个大小应都能正常显示
    * 缩小窗口导致一些控件的文字换行后可能有部分文字无法显示
    * 可以尝试使用 `Grid`、`WarpPanel`、`DockPanel`、`ViewBox` 等控件
* 尽量使用 PCL 自己的 `local:MyCard` 作为卡片最外层元素，使得主页与 PCL 整体协调
* 不要并排放置两个能够折叠的卡片，当折叠其一时会显得布局失衡
* 卡片间距不宜过小或过大，`10` 左右即可
* 控件之间应留有一定间隙 (`Margin`)
* 尽量不要卡片内套卡片

### 画面
* 整体使用到的颜色不宜过多。UI 控件（图片除外）尽量使用 `{DynamicResource ColorBrushX}` 笔刷
    * 在需要提醒注意的地方可以使用红色或其它颜色
    * 除主题色，其它颜色的饱和度不宜过高
* 尽量不要使用渐变色作为卡片背景
    * 不建议除设计目的（如红色提醒注意）等情况更改卡片背景颜色
* 应限制发光效果的使用
* 直角的控件建议打一定圆角以配合 PCL 整体圆角设计
* 半透明的控件要保证可读性

### 文字
* 字号适中，8 ~ 26 为宜
* 内容性主页正文颜色应使用黑色
* 除特殊情况（如 LOGO）外均应使用默认字体
* 可以使用 emoji 来作为图标
* 启用自动换行 `TextWrapping = "Wrap"`
* 注意可能会用到转义字符 `&amp` = `&`、`&quot;` = `"` 等等

### 主页
* 失败的绑定会对性能有影响，使用绑定时不要使绑定失败
* 使用 `Style` 会减少许多重复代码
* 使用 `local:MyImage` 控件代替 WPF 原生的 `Image` 控件
    * `local:MyImage` 控件提供了异步加载与加载失败后回调的图片链接，优化了主页性能

### 网络
* 挂载尽量使用中国大陆境内的服务，如服务器、COS、CDN
    * 境内服务可能需要备案
    * 临近地区的服务也可行，但要保证其访问性
    * 不要使用 Github Pages 等境内不方便访问的服务
* 需要使用大量图片建议使用图床等服务
* **发送 HTTP GET 请求后返回的一定要是 XAML 源代码，不能是 HTML**
    * 使用浏览器访问该网址一般会启动下载下载 .xaml 文件，若不是一般此处配置错误
    * 如果返回的是 HTML 的话，一般会报 “意外的DTD声明”、“无法创建未知类型"{...}html"”等
* 若需要，可以通过设置 index、301 302 重定向、服务器后端等方式将 `http://example.com/Custom.xaml` 改为 `http://example.com/`

#### 版本检查
> 你需要在该网址添加一个 **版本号文件**，文件内容和格式均随意。PCL 将在每次尝试加载主页前先下载该文件，如果版本号文件内容没有改变，则 **不会** 下载主页，而是使用本地缓存；仅当该文件内容改变时，才会重新下载主页。因此，你需要在 **每次主页内容改变时** ，任意修改该文件的内容。

* 版本号不应过长，否则会照成额外开销
* 版本号获取地址根据是不是 .xaml 结尾有这两种格式
    * 主页地址类似于 `http://example.com/home` 的会访问 `http://example.com/home/version` 获取版本号
    * 主页地址类似于 `http://example.com/home.xaml` 的会访问 `http://example.com/home.xaml.ini` 获取版本号

#### 流量过滤（可选）
* 限制 `Referer` 为 `*.pcl2.server` 和 `*.pcl2.open.server`
* 限制 `User Agent` 包含 `PCL2/*`
* 上述两条可以判断流量是否来源于 PCL（如果是针对性的攻击可能没有用）
* 访问频率：每分钟最多 2 次，每小时最多 10 次
    * 2 次是由于需要获取一次版本再可能获取一次主页内容决定的
    * 实际服务商可能采用 QPS 作为单位，基于相同道理，其值不能低于 2
* 可以设置一定带宽上限

## 其它要注意的
* [新] 自PCL 2.10.6 起 `IsSwaped` 重命名为 `IsSwapped`，虽然在新版本中保留了`IsSwaped`字段不会报错，但已无实际功能。需要判断 PCL 版本替换字段名称以保证主页功能正常运行。
* Image 控件**无法**用于展示 gif 动图，展示 gif 时图片会以静态呈现
* 使用 HTTPS 的 301 302 跳转时 `Header` 中不带 `Referer`，可能造成误拦，需要格外注意。
* 在 PCL 日志可以看到加载主页的时间开销

**Comment by wuliaodexiaoluo** [comment]
提一嘴，重复的主页除非比已存在的主页质量更高，否则不会加入预设

**Comment by MoYuan-CN** [comment]
在这里借一楼说一下
@LTCatt 是否有考虑单独开一个 `主页投稿` 分区，同时也能解决 #2528 下方楼太长不好找帖的问题？

**Reply by LTCatt** [reply]
整了

**Reply by LTCatt** [reply]
改了
